
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>Library API &#8212; CtrlF-TF 1.0b1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Command Line Interface" href="command_line_program.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="library-api">
<span id="id1"></span><h1>Library API<a class="headerlink" href="#library-api" title="Permalink to this heading">¶</a></h1>
<p>This page contains information about the classes in CtrlF-TF.</p>
<section id="alignparameters">
<h2>AlignParameters<a class="headerlink" href="#alignparameters" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="ctrlf_tf.AlignParameters">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ctrlf_tf.</span></span><span class="sig-name descname"><span class="pre">AlignParameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pwm_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pwm_file_format</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Tabular'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">core_start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">core_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">core_gaps</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">range_consensus</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palindrome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'1.0b1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kmer_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pwm_reverse_complement</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.AlignParameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Dataclass for input parameters to align k-mers to a model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kmer_file</strong> (<em>str</em>) – Pathway to a text file of kmers. Must have columns
representing: k-mer, reverse_complement k-mer, score where the score
will be used to determine the rank of the k-mer.</p></li>
<li><p><strong>pwm_file</strong> (<em>str</em>) – Pathway to a text file for a position weight matrix to use
as a model for alignment. Format is specified by the <em>pwm_file_format</em>
argument, which is “Tabular” by default. The pwm file is assumed to
define probabilities.</p></li>
<li><p><strong>pwm_file_format</strong> (<em>str</em>) – Indicates if the position weight matrix specified
in the pwm_file param is in a tabular or meme format by specifying
either <em>“Tabular”</em> or <em>“MEME”</em>.</p></li>
<li><p><strong>core_start</strong> (<em>int</em>) – 1-based, inclusive position for the start of the core.
The core defines the positions that must be described by kmers to call
a site.</p></li>
<li><p><strong>core_end</strong> (<em>int</em>) – 1-based inclusive position for the end of the core. The
core defines the positions that must be described by kmers to call a
site.</p></li>
<li><p><strong>core_gaps</strong> (<em>Iterable</em>) – 1-based positions relative to core start that define
sections within a core that do not need to be described. These
positions will be equiprobable during alignment.</p></li>
<li><p><strong>range_consensus</strong> (<em>str</em>) – A kmer input that defines core_start, core_end, and
core_gaps by aligning the kmer to the PWM based on the maximum score
and selecting the described positions as a core. Using a ‘.’ in the
kmer input will make that position a core_gap.</p></li>
<li><p><strong>gap_limit</strong> (<em>int</em>) – Limit on the number of gaps to allow kmers to have. Must
be 0 or greater.</p></li>
<li><p><strong>threshold</strong> (<em>float</em>) – A score to filter aligned kmers from. The column used to
filter is determined by <em>threshold_column</em>.</p></li>
<li><p><strong>threshold_column</strong> (<em>str</em>) – Column name in the kmer file to filter for kmers
above a threshold determined by the <em>threshold</em> parameter.</p></li>
<li><p><strong>palindrome</strong> (<em>bool</em>) – If <em>true</em>, both orientations of kmers are aligned to the
core pwm model and orientations in called sites will me ‘.’. If <em>false</em>
then the maximum scoring kmer of the two orientations will be used.
Sites will be called with a ‘+’ or ‘-’ orientation.</p></li>
<li><p><strong>version</strong> (<em>str</em>) – Version of ctrlf_tf used in alignment</p></li>
<li><p><strong>pwm_reverse_complement</strong> (<em>bool</em>) – If <em>true</em>, the PWM orientation is the
reverse complement of the input PWM.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.AlignParameters.save_parameters">
<span class="sig-name descname"><span class="pre">save_parameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'w'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.AlignParameters.save_parameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves the parameters to a file.</p>
</dd></dl>

</dd></dl>

</section>
<section id="bedtuple">
<h2>BedTuple<a class="headerlink" href="#bedtuple" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="ctrlf_tf.site_call_utils.BedTuple">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ctrlf_tf.site_call_utils.</span></span><span class="sig-name descname"><span class="pre">BedTuple</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.site_call_utils.BedTuple" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</section>
<section id="classifiedsequences">
<h2>ClassifiedSequences<a class="headerlink" href="#classifiedsequences" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="ctrlf_tf.ClassifiedSequences">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ctrlf_tf.</span></span><span class="sig-name descname"><span class="pre">ClassifiedSequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">classified_dataframe</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negative_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ThresholdTuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positive_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ThresholdTuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'1.0b1'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.ClassifiedSequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Classify de Bruijn sequences.</p>
<p>Takes an input dataframe in the format (with or without headers):</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Value</p></th>
<th class="head"><p>Sequence</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>50000</p></td>
<td><p>CAATCG…</p></td>
</tr>
<tr class="row-odd"><td><p>49554</p></td>
<td><p>ATCGAT…</p></td>
</tr>
<tr class="row-even"><td><p>…</p></td>
<td><p>…</p></td>
</tr>
</tbody>
</table>
<p>and creates a classified dataframe in the format:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Value</p></th>
<th class="head"><p>Sequence</p></th>
<th class="head"><p>Group</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>50000</p></td>
<td><p>CAATCG…</p></td>
<td><p>‘+’</p></td>
</tr>
<tr class="row-odd"><td><p>49554</p></td>
<td><p>ATCGAT…</p></td>
<td><p>‘-’</p></td>
</tr>
<tr class="row-even"><td><p>…</p></td>
<td><p>…</p></td>
<td><p>…</p></td>
</tr>
</tbody>
</table>
<p>The Group column indicates if it is a bound sequence, not bound or
ambiguous (+/-/.).</p>
<p>Meta data is accessed with the following attributes:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>negative_threshold: ThresholdTuple that describes the definition and</dt><dd><p>value of the negative group “-”</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>positive_threshold: ThresholdTuple that described the definition and</dt><dd><p>value of the positive group “+”</p>
</dd>
</dl>
</li>
<li><p>version: Version of ctrlf_tf used to generate the classification</p></li>
</ol>
<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.ClassifiedSequences.classify_from_dataframe">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">classify_from_dataframe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'kde_z4'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_negative</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_positive</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ln_transform</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.ClassifiedSequences.classify_from_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Classifies a set of de Bruijn sequences using a given method.</p>
<p>Factory method to create a ClassifiedDeBruijn object from a pandas
DataFrame. The dataframe must have columns in the following format
with or without headers:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Value</p></th>
<th class="head"><p>Sequence</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>50000</p></td>
<td><p>CAATCG…</p></td>
</tr>
<tr class="row-odd"><td><p>49554</p></td>
<td><p>ATCGAT…</p></td>
</tr>
<tr class="row-even"><td><p>…</p></td>
<td><p>…</p></td>
</tr>
</tbody>
</table>
<p>By default, classification is done using the kde_z4 method. This
calculates the Kernal Density Estimate (Gaussian) of the
distribution and selects the maximum result of each value as an
input. Then the distance between the minimum value and this value
is multiplied by 2 and that is the kde threshold. This is compared
to a modified z-score based on medians of 4 (assuming a Gaussian
distribution). The negative and positive thresholds are determined
by the minimum and maximum of these results respectively.</p>
<p>Otherwise, classification can be done using the “z-score” method.
This calculated the modified z-score for a given negative and
positive value as the negative and positive thresholds respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataframe</strong> (<em>Pandas DataFrame</em>) – Dataframe of the de Bruijn sequence information</p></li>
<li><p><strong>method</strong> (<em>str</em>) – Method of classification (kde_z4 or z-score)</p></li>
<li><p><strong>z_negative</strong> (<em>float</em>) – Negative z-score value if using z-score method
(default = 3)</p></li>
<li><p><strong>z_positive</strong> (<em>float</em>) – Positive z-score value if using z-score method
(default = 4)</p></li>
<li><p><strong>sequence_start</strong> (<em>int</em>) – Optional start index for the de Bruijn sequence
(default None)</p></li>
<li><p><strong>sequence_end</strong> (<em>int</em>) – Optional end index for the de Bruijn sequence
(default None)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>ClassifiedDeBruijn Object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.ClassifiedSequences.load_from_file">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_from_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.ClassifiedSequences.load_from_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize a ClassifiedDeBruijn object from a file.</p>
<p>Initialize a CLassifiedDeBruijn object from a text file. The file
has to be in the same format as the output from the save_to_file
method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file_path</strong> – File path to load the ClassifiedDeBruijn from</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>ClassifiedDeBruijn object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.ClassifiedSequences.save_to_file">
<span class="sig-name descname"><span class="pre">save_to_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.ClassifiedSequences.save_to_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Save classification data to a file in tabular format with meta data.</p>
<p>Saves the information from the class attributes to a text file output.
This file can be loaded by load_from_file to initialize a
ClassifiedDeBruijn object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file_path</strong> – Relative path to save the attribute information to</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="ctrlf">
<h2>CtrlF<a class="headerlink" href="#ctrlf" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="ctrlf_tf.CtrlF">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ctrlf_tf.</span></span><span class="sig-name descname"><span class="pre">CtrlF</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">core_positions</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aligned_kmer_dataframe</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palindrome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pwm</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">ndarray</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kmer_dataframe</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compiled_site_dataframe</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">abs_core_start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">abs_core_end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.CtrlF" title="Permalink to this definition">¶</a></dt>
<dd><p>Class used to align, compile, and call sites.</p>
<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.CtrlF.call_sites">
<span class="sig-name descname"><span class="pre">call_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fixed_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.CtrlF.call_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of SiteTuples from an input sequence.</p>
<p>Given a sequence, returns a list of SiteTuples for each called site.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<em>str</em>) – Input DNA sequence</p></li>
<li><p><strong>fixed_length</strong> (<em>bool</em>) – Search mode assumes a fixed model length</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of SiteTuples</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.CtrlF.call_sites_as_bed">
<span class="sig-name descname"><span class="pre">call_sites_as_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fixed_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.CtrlF.call_sites_as_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Call sites in BED format.</p>
<p>Given a seuqence, chromosome, and chromosome_start information, returns
called sites as a list of BedTuples</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence</strong> (<em>str</em>) – Input DNA sequence</p></li>
<li><p><strong>chromosome</strong> (<em>str</em>) – Chromosome label</p></li>
<li><p><strong>chromosome_start</strong> (<em>int</em>) – Start position of the input sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of BedTuples</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.CtrlF.call_sites_from_fasta">
<span class="sig-name descname"><span class="pre">call_sites_from_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fasta_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genomic_label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.CtrlF.call_sites_from_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a fasta file input, calls sites in BED format.</p>
<p>With a fasta file as input, calls sites from every sequence. Returns
output in BED format. By default, the chromosome is the full header and
the start and end are relatve to sequence length. With the genomic
label parameter set to True, the header is parsed if given in the
format: Chromosome:start-end. By default the method returns a pandas
DataFrame, but with the to_file parameter set to a file location, the
function will write output to the location. This differs in that called
site data is not kept in memory and is written in groups of sites
called per sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fasta_file</strong> (<em>str</em>) – A fasta file input of sequences</p></li>
<li><p><strong>genomic_label</strong> (<em>bool</em>) – Flag for if the headers are genomic coordinates</p></li>
<li><p><strong>to_file</strong> (<em>str</em>) – File location output. If specified, outputs sites
for each sequence one set at a time.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>By default a Pandas DataFrame of called sites in bed format but if to_file is specified, writes the results to a file in bed format</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.CtrlF.from_alignment_file">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_alignment_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.CtrlF.from_alignment_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct class using a previous AlignedKmers output file.</p>
<p>Parses the information saved from using the method
<em>.save_alignment</em> to make a new class instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file_path</strong> (<em>str</em>) – Path to saved alignment file</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Class instance with data from the alignment file</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.CtrlF.from_compiled_sites">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_compiled_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.CtrlF.from_compiled_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct class from <em>save_compiled_sites()</em> output.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file_path</strong> (<em>str</em>) – File location of compiled sites.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.CtrlF.from_parameters">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_parameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameters</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#ctrlf_tf.AlignParameters" title="ctrlf_tf.ctrlf_core.AlignParameters"><span class="pre">AlignParameters</span></a></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.CtrlF.from_parameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Construcor using an AlignedParameters object.</p>
<p>A factory constructor that aligned kmers to a PWM model using
parameters defined in a <em>AlignParameters</em> class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.CtrlF.save_alignment">
<span class="sig-name descname"><span class="pre">save_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.CtrlF.save_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Save alignment data to stdout or a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>location</strong> (<em>str</em>) – Output location, if None output will be stdout</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.CtrlF.save_compiled_sites">
<span class="sig-name descname"><span class="pre">save_compiled_sites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.CtrlF.save_compiled_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Saves compiled sites as a table to a file or stdout.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output</strong> (<em>str</em>) – Output location (default = stdout)</p></li>
<li><p><strong>minimal</strong> – If <em>true</em>, removes column showing which kmer indexes
were used to generate the solution.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="optimize">
<h2>Optimize<a class="headerlink" href="#optimize" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="ctrlf_tf.Optimize">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ctrlf_tf.</span></span><span class="sig-name descname"><span class="pre">Optimize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">align_parameters</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#ctrlf_tf.AlignParameters" title="ctrlf_tf.ctrlf_core.AlignParameters"><span class="pre">AlignParameters</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">classified_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fpr_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'1.0b1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameter_dataframe</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tpr_fpr_dictionary</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">dict</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimal_parameters</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#ctrlf_tf.AlignParameters" title="ctrlf_tf.ctrlf_core.AlignParameters"><span class="pre">AlignParameters</span></a><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_thresholds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{0:</span> <span class="pre">0.35,</span> <span class="pre">1:</span> <span class="pre">0.4,</span> <span class="pre">2:</span> <span class="pre">0.43}</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.Optimize" title="Permalink to this definition">¶</a></dt>
<dd><p>Optimize AlignParameters using classified de Bruijn sequences.</p>
<p>Optimizes AlignParameters using de Bruijn sequence data classified
into positive and negative probes. Optimization based on AUROC at a target
false positive rate (default = 0.01).</p>
<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.Optimize.load_from_file">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load_from_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.Optimize.load_from_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Load an Optimize class from a saved text file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ctrlf_tf.Optimize.save_to_file">
<span class="sig-name descname"><span class="pre">save_to_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.Optimize.save_to_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Save optimized parameter information to a text file.</p>
<p>The saved file contains the attributes of the Optimize object,
including:
1) Inital AlignParameter
2) Input Classified deBruijn DataFrame
3) DataFrame of all benchmarked parameters up to the optimal one
4) DataFrame of all TPR and FPR results</p>
<p>The save file generated from this method can be used by the
load_from_file method to create a new Optimize object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file_path</strong> – File path to save the attribute information</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="sitetuple">
<h2>SiteTuple<a class="headerlink" href="#sitetuple" title="Permalink to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="ctrlf_tf.site_call_utils.SiteTuple">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ctrlf_tf.site_call_utils.</span></span><span class="sig-name descname"><span class="pre">SiteTuple</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orientation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#ctrlf_tf.site_call_utils.SiteTuple" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">CtrlF-TF</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="optimization.html">Optimization</a></li>
<li class="toctree-l1"><a class="reference internal" href="command_line_program.html">Command Line Interface</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Library API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#alignparameters">AlignParameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bedtuple">BedTuple</a></li>
<li class="toctree-l2"><a class="reference internal" href="#classifiedsequences">ClassifiedSequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ctrlf">CtrlF</a></li>
<li class="toctree-l2"><a class="reference internal" href="#optimize">Optimize</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sitetuple">SiteTuple</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="command_line_program.html" title="previous chapter">Command Line Interface</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2022, Zachery Mielko.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/library_api.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>